### PR TITLE
Ai attack/repair basic fix

### PR DESCRIPTION
**Pull request by dnqbob**
@reaperrr 
@abcdefg30

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs [hunk 1]
```diff
@@ -199,12 +177,14 @@ public void Tick(Squad owner)
 				if (BusyAttack(a))
 					continue;
 
-				if (!ReloadsAutomatically(a))
+				var ammoPools = a.TraitsImplementing<AmmoPool>().ToArray();
+
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs]: This empty line can be removed.

File: OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs [hunk 2]
```diff
@@ -120,5 +120,24 @@ protected static bool IsRearming(Actor a)
...
+		}
+
+		protected static bool ReloadsAutomatically(IEnumerable<AmmoPool> ammoPools, Rearmable rearmable)
+		{
+			if (rearmable == null)
+				return true;
+
+			return ammoPools.All(ap => !rearmable.Info.AmmoPools.Contains(ap.Info.Name));
+		}
+
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs:150]: This empty line must be removed.

File: OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs [hunk 3]
```diff
@@ -230,7 +210,8 @@ public void Tick(Squad owner)
 
 			foreach (var a in owner.Units)
 			{
-				if (!ReloadsAutomatically(a) && !FullAmmo(a))
+				var ammoPool = a.TraitsImplementing<AmmoPool>().ToArray();
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs]: ammoPool**s**.

Also, since we're dropping the `.All` in the methods the `.ToArray()` can go away.

**Comment by dnqbob** [comment]
So after you check and merge this, I will open several other prs based on bleed.

File: OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs [hunk 4]
```diff
@@ -120,5 +120,24 @@ protected static bool IsRearming(Actor a)
 
 			return false;
 		}
+
+		protected static bool FullAmmo(IEnumerable<AmmoPool> ammoPools)
+		{
+			return ammoPools.All(x => x.HasFullAmmo);
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs]: This means we always return `false` if any pool doesn't have full ammo, so it would be more performant to do
```
foreach (var ap in ammoPools)
	if (!ap.HasFullAmmo)
		return false;

return true;
```

File: OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs [hunk 5]
```diff
@@ -120,5 +120,24 @@ protected static bool IsRearming(Actor a)
...
+		}
+
+		protected static bool HasAmmo(IEnumerable<AmmoPool> ammoPools)
+		{
+			return ammoPools.All(x => x.HasAmmo);
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs]: Same as above.

File: OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs [hunk 6]
```diff
@@ -120,5 +120,24 @@ protected static bool IsRearming(Actor a)
...
+		}
+
+		protected static bool ReloadsAutomatically(IEnumerable<AmmoPool> ammoPools, Rearmable rearmable)
+		{
+			if (rearmable == null)
+				return true;
+
+			return ammoPools.All(ap => !rearmable.Info.AmmoPools.Contains(ap.Info.Name));
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs]: Same here:
```
foreach (var ap in ammoPools)
	if (!rearmable.Info.AmmoPools.Contains(ap.Info.Name))
		return false;

return true;
```

File: OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs [hunk 7]
```diff
@@ -199,12 +177,14 @@ public void Tick(Squad owner)
 				if (BusyAttack(a))
 					continue;
 
-				if (!ReloadsAutomatically(a))
+				var ammoPools = a.TraitsImplementing<AmmoPool>().ToArray();
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs:180]: Since we're dropping the `.All` in the methods, the `.ToArray()` can go away.

**Review by reaperrr — DISMISSED** [review]

File: OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs (hunk 1 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs]: checkmark

File: OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs (hunk 2 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs:150]: checkmark

**Comment by deleted-user-1** [comment]
Please rework the commit messages. I should know from reading the commit messages what type of changes you did and for what reason.

`basic fix on AI state related` can mean anything about AI state. If the types of changes in this commit (visibility/rearming) have nothing in common and that's the reason for the unspecific commit message, consider splitting them into individual commits.

`AirState basic fix` also tells me nothing useful. 

`"IsRearm" fix and name change`

What type of fix?

File: OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs (hunk 3 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs]: checkmark

File: OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs (hunk 4 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs]: checkmark

File: OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs (hunk 5 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs]: checkmark

File: OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs (hunk 6 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs]: checkmark

**Comment by dnqbob** [comment]
commit message fixed

**Comment by deleted-user-1** [comment]
I still don't understand why the `"FindEnemy" functions ignore hidden actors` commit has to care about rearming state or what the "IsRearm" fix is about. Not an issue though, thanks for applying the changes.

**Comment by dnqbob** [comment]
> I still don't understand why the "FindEnemy" functions ignore hidden actors commit has to care about rearming state or what the "IsRearm" fix is about. Not an issue though, thanks for applying the changes.

ummm... because title is short so I have to discribe it within words, probably crude. Further detail can be seen in files so I think it is unecessary here. 

Thanks your advice anyway.

File: OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs [hunk 8]
```diff
@@ -160,12 +175,14 @@ void IBotTick.BotTick(IBot bot)
...
-			return World.Actors.Where(IsEnemyUnit).ClosestTo(pos);
+			var units = World.Actors.Where(IsEnemyUnit);
+			return units.Where(IsNotHiddenUnit).ClosestTo(pos) ?? units.ClosestTo(pos);
 		}
 
 		internal Actor FindClosestEnemy(WPos pos, WDist radius)
 		{
-			return World.FindActorsInCircle(pos, radius).Where(IsEnemyUnit).ClosestTo(pos);
+			var units = World.FindActorsInCircle(pos, radius).Where(IsEnemyUnit);
+			return units.Where(IsNotHiddenUnit).ClosestTo(pos);
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs]: This is only a question for now, not a change request (yet): Is there a reason why this doesn't do the same ` ?? units.ClosestTo(pos)` as the overload above?

File: OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs (hunk 2 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs:150]: I guess I remove it here?

**Comment by reaperrr** [comment]
> I still don't understand why the "FindEnemy" functions ignore hidden actors commit has to care about rearming state

I agree It would be better if the `IsRearm` moving to `StateBase` from commit 1 was merged into commit 2, yes.

>  or what the "IsRearm" fix is about

I guess it's about the type lookup no longer being pulled out into a var (since that was redundant), but I agree it's so minor that it would be less confusing if the 'fix' part was removed from the commit message.

File: OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs (hunk 8 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs]: Because find actors in `FindClosestEnemy(WPos pos, WDist radius)` is often used for target scanning within an area. If we apply `?? units.ClosestTo(pos)` here then AI will try to attack it, when all other not-hidden units are killed (but stucked because it cannot be attacked when cloaked). 

`FindClosestEnemy(WPos pos)` is often used for approaching target for AI squad. So `?? units.ClosestTo(pos)` here means if AI really cannot find any enemy **on map** that not cloaked, they will finally try to approach the cloacked one.

**Comment by dnqbob** [comment]
Ah, my mistakes here. I use this branch as base branch yesterday but never check the commit problem.

**Comment by dnqbob** [comment]
.... please give me sometime here, It seems I mistake something again for another place.

**Comment by dnqbob** [comment]
Need to learn git more for me.

File: OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs (hunk 8 above)
Comment (reaperrr) [reply OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs]: In that case it should be possible to combine the two `Where()` though, since both need to be checked anyway. Then we also wouldn't need the `var` and could just return
```
World.FindActorsInCircle(pos, radius).Where(a => IsEnemyUnit(a) && IsNotHiddenUnit(a)).ClosestTo(pos)
```
which should be a bit more performance-efficient.

Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs]: Fixed

**Review by reaperrr — DISMISSED** [review]
Code looks good to me now, didn't spot any issues in a 4-AI RA game, either.

**Comment by dnqbob** [comment]
dude,  you can also try lay some mines in front of AI's door,  and see if they will stuck when attack.  

There are still some lines needs to be fixed in states because they directly use `isEnemyUnit` to detect enemy,  **if we want ignore hidden enemy works normally, but not for now**. Unfixed lines will let AttackMove state switch to Attack state when encounter the hidden unit, although they will begin to search and attack visible one during Attack state.

File: OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs [hunk 9]
```diff
@@ -220,12 +177,13 @@ public void Tick(Squad owner)
 				if (BusyAttack(a))
 					continue;
 
-				if (!ReloadsAutomatically(a))
+				var ammoPools = a.TraitsImplementing<AmmoPool>();
+				if (!ReloadsAutomatically(ammoPools, a.TraitOrDefault<Rearmable>()))
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs:181]: I must have missed this during all the back and forth. Can someone elaborate why we're suddenly doing the trait lookups outside of the method (= high coupling)? Since we're not `ToArray`ing or `ToList`ing we'll still enumerate twice worst case, so there should basically be no difference between doing the lookups here and doing it inside the method (except coupling).

Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs:181]: ummmm.. yeah I also said readability is down here in exchange of the performance.

Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs:181]: But it will require more thinking on probably change all those functions back for readability and low coupling, since they are all about the ammo related.

Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs:181]: My point was that this doesn't look like it gives us any performance gains.

Comment (reaperrr) [reply OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs:181]: The idea was that both `ReloadsAutomatically` and `HasAmmo()` below were looking up `a.TraitsImplementing<AmmoPool>();` before, so I was assuming that caching it once, passing it through as argument and then running a foreach on them would save at least one of the (up to) two look-ups as well as the additional enumerations.

I'm not that firm on the topic though, if you say passing them as array would be more efficient we should do that, of course.

Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs:181]: Changed to `ToArray`

File: OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs (hunk 8 above)
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs]: Why was this change reverted again?

File: OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs [hunk 10]
```diff
@@ -251,9 +209,10 @@ public void Tick(Squad owner)
 
 			foreach (var a in owner.Units)
 			{
-				if (!ReloadsAutomatically(a) && !FullAmmo(a))
+				var ammoPools = a.TraitsImplementing<AmmoPool>();
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs]: This has the same problem.

Fwiw, I'm still not a fan of the approach, but if you both think it is worth it we can keep it.

File: OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs (hunk 8 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs]: Oh...ummm for fixing the first commit I drop the second then copy a back up one

File: OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs (hunk 10 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs]: Ummm, I quite don't know how many performance it will bring. But AI tick can sometimes lag the game (mainly pathfinding, though). 

I just finished a GroundAttackMove state fix up to #17327(thanks tovl's idea), in my test map (a lot of twisted road) the game will lag everytime when AI attackmove. AI airforce can also lags for some unknown reasons when there are many aircrafts (they don't even do complex pathfinding). I don't know will improvement of some places can help on the lags caused by AI.

File: OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs (hunk 8 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs]: Should be fixed

File: OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs (hunk 10 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs]: should be fixed

File: OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs [hunk 11]
```diff
@@ -126,6 +126,21 @@ public bool IsEnemyUnit(Actor a)
 				&& !a.GetEnabledTargetTypes().IsEmpty;
 		}
 
+		public bool IsNotHiddenUnit(Actor a)
+		{
+			var hasModifier = false;
+			var visModifiers = a.TraitsImplementing<IVisibilityModifier>().ToArray();
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs]: Now this `ToArray` is one too many again, sorry.

**Review by abcdefg30 — COMMENTED** [review]
Some explanation to when we should use `ToArray`:
`TraitsImplementing` returns an `IEnumerable`. That means it does basically nothing until you start looping over the enumerable. While looping the next element is determined/fetched for each loop run individually. When we loop a second time over the entire enumerable this happens again.
Now, calling `ToArray` or `ToList` forces the enumerable to do this right away; to determine all elements and save them somewhere to memory. Iterating over that list/array will now be much faster than fetching each element again like before.
Tl;dr: Using `IEnumerable` == do the work when we actually run through a loop or similar, while `ToList`ing/`ToArray`ing == do the work right away but be faster when we loop later.

If we only ever loop once directly after creating the enumerable you should usually not use `ToArray` or `ToList` since that has the overhead of allocating memory (and in theory a second iteration: The first is the iteration when using ToArray, and the second for looping over that array).
If you loop several times you should use `ToArray` or `ToList` since the array/list enumerators are much faster than evaluating the `IEnumerable` again each time.

**Comment by abcdefg30** [comment]
Some links if you are interested:
https://stackoverflow.com/questions/23536541/performance-between-iterating-through-ienumerablet-and-listt
https://stackoverflow.com/questions/3628425/ienumerable-vs-list-what-to-use-how-do-they-work

**Comment by dnqbob** [comment]
> Some links if you are interested:
> https://stackoverflow.com/questions/23536541/performance-between-iterating-through-ienumerablet-and-listt
> https://stackoverflow.com/questions/3628425/ienumerable-vs-list-what-to-use-how-do-they-work

Marked.

File: OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs (hunk 11 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs]: It is OK here. I also update my back up.

**Comment by dnqbob** [comment]
Does this mean, the `IEnumerable` is just a placeholder which later may be used (in Linq), every time it is being used will result in an extra process of converting to list/array?

```
// IEn is a Linq based "IEnumerable"
X1 = IEn.Select(...).Where(...);
X2 = X1.Select(...).Where(...);

Y1 = IEn.Select(...).Where(...).ToArray();
```
 X1 and X2 are nothing but placeholders, Y is actually an array?

**Comment by abcdefg30** [comment]
> X1 and X2 are nothing but placeholders, Y is actually an array?

You can put it like that, yes.

> Does this mean, the `IEnumerable` is just a placeholder which later may be used (in Linq), every time it is being used will result in an extra process of converting to list/array?

"Placeholder which we use later" yes. But it is not converted to a list/an array. I think you got the idea though: Every time it is used we will need to fetch all the elements again. (This just doesn't happen all at once, but after each time we ran through a `foreach` body the next element is fetched. So you basically do "fetch next element - run loop body - fetch next element - run loop body - ..." afaik.)

**Review by abcdefg30 — APPROVED** [review]
Ok, lgtm.

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/92d5fa562493b758f23cdc97bdc22e04c8d4459f...114f2b372a2f419b4943d8de354f19a7cf17d5a1)
